src/main/java/org/folio/bulkops/processor/marc/MarcFlowCommitProcessor.java

```diff
@@ -0,0 +1,179 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class MarcFlowCommitProcessor {
+  private final RemoteFileSystemClient remoteFileSystemClient;
+  private final ObjectMapper objectMapper;
+
+  public void processMarcFlowCommitResult(BulkOperation bulkOperation) {
+    var csvHrids = getUpdatedInventoryInstanceHrids(bulkOperation);
+    var marcHrids = getUpdatedMarcInstanceHrids(bulkOperation);
+
+    var updatedHrids = new HashSet<>(csvHrids);
+    updatedHrids.addAll(marcHrids);
+    bulkOperation.setCommittedNumOfRecords(updatedHrids.size());
+
+    enrichCommittedCsvWithUpdatedMarcRecords(bulkOperation, csvHrids, marcHrids);
+    enrichCommittedMarcWithUpdatedInventoryRecords(bulkOperation, csvHrids, marcHrids);
+  }
+
+  public void enrichCommittedCsvWithUpdatedMarcRecords(BulkOperation bulkOperation, List<String> csvHrids, List<String> marcHrids) {
+    var hrids = CollectionUtils.subtract(marcHrids, csvHrids);
+    if (!hrids.isEmpty() && nonNull(bulkOperation.getLinkToMatchedRecordsJsonFile())) {
+      var dirName = isNull(bulkOperation.getLinkToCommittedRecordsCsvFile()) ? bulkOperation.getId().toString() : ENRICHED_PREFIX + bulkOperation.getId();
+      var committedCsvFileName = CHANGED_CSV_PATH_TEMPLATE.formatted(dirName, LocalDate.now(), getBaseName(bulkOperation.getLinkToTriggeringCsvFile()));
+      try (var matchedFileReader = new InputStreamReader(new BufferedInputStream(remoteFileSystemClient.get(bulkOperation.getLinkToMatchedRecordsJsonFile())));
+           var writer = isNull(bulkOperation.getLinkToCommittedRecordsCsvFile()) ?
+             remoteFileSystemClient.writer(committedCsvFileName) : new StringWriter()) {
+        var matchedFileParser = new JsonFactory().createParser(matchedFileReader);
+        var matchedFileIterator = objectMapper.readValues(matchedFileParser, ExtendedInstance.class);
+        var csvWriter = new BulkOperationsEntityCsvWriter(writer, Instance.class);
+        while (matchedFileIterator.hasNext()) {
+          var instance = matchedFileIterator.next().getEntity();
+          if (hrids.contains(instance.getHrid())) {
+            writeBeanToCsv(csvWriter, instance);
+          }
+        }
+        if (nonNull(bulkOperation.getLinkToCommittedRecordsCsvFile())) {
+          var appendedCsvRecords = writer.toString();
+          appendedCsvRecords = appendedCsvRecords.substring(appendedCsvRecords.indexOf(LINE_BREAK) + 1);
+          var appendedCsvStream = new SequenceInputStream(
+            remoteFileSystemClient.get(bulkOperation.getLinkToCommittedRecordsCsvFile()),
+            new ByteArrayInputStream(appendedCsvRecords.getBytes()));
+          remoteFileSystemClient.put(appendedCsvStream, committedCsvFileName);
+        }
+      } catch (Exception e) {
+        log.error("Failed to enrich csv file", e);
+      } finally {
+        if (nonNull(bulkOperation.getLinkToCommittedRecordsCsvFile())) {
```

Comment:
This can be moved from finally to try section. This will allow to handle potential S3 issues in the same manner.